core-processor/src/ext.rs

```diff
@@ -872,19 +876,16 @@ impl Ext {
         pages: WasmPage,
         mem: &mut impl Memory,
     ) -> Result<WasmPage, ProcessorAllocError> {
-        // Charge gas for memory grow.
-        // TODO: move charging for grow inside alloc function, so that we can skip refunding #2337
-        let gas_refunder =
-            GasRefunder::charge_if_enough(self, self.context.page_costs.mem_grow.calc(pages))?;
-        let AllocInfo { page, not_grown } =
-            self.context.allocations_context.alloc::<G>(pages, mem)?;
-
-        // Refund gas if memory has not been grown to `pages` size.
-        gas_refunder
-            .refund(self, self.context.page_costs.mem_grow.calc(not_grown))
-            .unwrap_or_else(|err| unreachable!("Alloc implementation error: {err}"))?;
-
-        Ok(page)
+        self.context
+            .allocations_context
+            .alloc::<G>(pages, mem, |pages| {
+                Ext::charge_gas_if_enough(
+                    &mut self.context.gas_counter,
+                    &mut self.context.gas_allowance_counter,
+                    self.context.page_costs.mem_grow.calc(pages),
```

Comment:
it's amount of gas to be charged for growed `pages` amount of pages, but this count is differ in case. provide tests for it as well please